### DISCUSSION TITLE
how to set uvc camera resolution?

### DISCUSSION THREAD

**Discussion by Alvazz** [lfroms/viewfinder#18] [Q&A]
Nice app, but how do I set the resolution of the uvc camera?

**Comment by lfroms** [comment]
Converted to a Discussion since that's more suitable for questions.

According to section 4.2.2.4.3 of the UVC 1.5 Class Specification, the video resolution control is supported, but only for the Encoding Unit for cameras that have built-in encoders.

It is possible to add full support for controlling the Encoding Unit to Viewfinder, but most cameras will likely not support it, so it may not be useful.  For the most cameras, the software consuming the video feed will be responsible for setting the supported input resolution. My understanding is that the camera can report its supported resolutions to whatever software is being used, then the software selects an option or your own setting (if supported).  And I suspect overriding the resolution with AVFoundation would produce a conflict with Viewfinder and the application receiving the video feed.

Some UVC cameras may have vendor-specific Extension Units available to control resolution, but support largely depends on vendor.  Would need to take a closer look, but it doesn't sound like this could be a universal feature.

I'm not an expert on UVC, so if you have any resources that hint at this being possible (I'm doubtful at this point) I'd be happy to take a look!

**Reply by esternin** [reply]
If you read tlc/tk, perhaps my simple [tkCam](https://github.com/esternin/tkCam) app will answer the question.

I use vanillawish as the tcl/tk interpreter that comes with tlcuvc implementation built-in, so I can focus on functionality.

Most of the modern webcams do stream compression in-camera, so no need to capture individual frames as images and process them in software, just pipe the stream from the usb camera channel to an open file channel.  tlcuvc has all the hooks for it, formats, exposure and focus controls - if supported by the cameras, and capturing/streaming the feed.  You can re-code the file later using ffmpeg if a different format is required.
